ports/stm32f4/common-hal/busio/UART.c

```diff
@@ -36,45 +36,455 @@
+
+    return len;
+}
+
+void HAL_UART_RxCpltCallback(UART_HandleTypeDef *handle)
+{
+    for(int i=0; i<7; i++) {
+        //get context pointer and cast it as struct pointer
+        busio_uart_obj_t * context = (busio_uart_obj_t *)MP_STATE_PORT(cpy_uart_obj_all)[i];
+        if(handle == &context->handle) {
+            //check if transaction is ongoing
+            if((HAL_UART_GetState(handle) & HAL_UART_STATE_BUSY_RX) == HAL_UART_STATE_BUSY_RX) {
+                return;
+            }
+            ringbuf_put_n(&context->rbuf, &context->rx_char, 1);
+            errflag = HAL_UART_Receive_IT(handle, &context->rx_char, 1);
+            
+            return;
+        }
+    }
+}
+
+void HAL_UART_ErrorCallback(UART_HandleTypeDef *UartHandle)
+{
+    if (__HAL_UART_GET_FLAG(UartHandle, UART_FLAG_PE) != RESET) {
+        __HAL_UART_CLEAR_PEFLAG(UartHandle);
+    } else if (__HAL_UART_GET_FLAG(UartHandle, UART_FLAG_FE) != RESET) {
+        __HAL_UART_CLEAR_FEFLAG(UartHandle);
+    } else if (__HAL_UART_GET_FLAG(UartHandle, UART_FLAG_NE) != RESET) {
+        __HAL_UART_CLEAR_NEFLAG(UartHandle);
+    } else if (__HAL_UART_GET_FLAG(UartHandle, UART_FLAG_ORE) != RESET) {
+        __HAL_UART_CLEAR_OREFLAG(UartHandle);
+    }
+    //restart serial read after an error
+    for(int i=0; i<7; i++) {
+        busio_uart_obj_t * context = (busio_uart_obj_t *)MP_STATE_PORT(cpy_uart_obj_all)[i];
+        if(UartHandle == &context->handle) {
+            HAL_UART_Receive_IT(UartHandle, &context->rx_char, 1);
+            return;
+        }
+    }
+
 }
 
 uint32_t common_hal_busio_uart_get_baudrate(busio_uart_obj_t *self) {
-    return 0;
+    return self->baudrate;
 }
 
 void common_hal_busio_uart_set_baudrate(busio_uart_obj_t *self, uint32_t baudrate) {
+    //Don't reset if it's the same value
+    if (baudrate == self->baudrate) return;
+
+    //Otherwise de-init and set new rate
+    if(HAL_UART_DeInit(&self->handle) != HAL_OK) {
+        mp_raise_ValueError(translate("UART De-init error"));
+    }
+    self->handle.Init.BaudRate = baudrate;
+    if(HAL_UART_Init(&self->handle) != HAL_OK) {
+        mp_raise_ValueError(translate("UART Re-init error"));
+    }
+
+    self->baudrate = baudrate;
 }
 
 uint32_t common_hal_busio_uart_rx_characters_available(busio_uart_obj_t *self) {
-    return 0;
+    return ringbuf_count(&self->rbuf);
 }
 
 void common_hal_busio_uart_clear_rx_buffer(busio_uart_obj_t *self) {
+    // Halt reception
+    HAL_NVIC_DisableIRQ(self->irq);
+    ringbuf_clear(&self->rbuf);
+    HAL_NVIC_EnableIRQ(self->irq);
 }
 
 bool common_hal_busio_uart_ready_to_tx(busio_uart_obj_t *self) {
-    return 0;
+    return __HAL_UART_GET_FLAG(&self->handle,UART_FLAG_TXE);
+}
+
+STATIC void call_hal_irq(int uart_num) {
+    //Create casted context pointer
+    busio_uart_obj_t * context = (busio_uart_obj_t *)MP_STATE_PORT(cpy_uart_obj_all)[uart_num-1];
+    if(context != NULL) {
+        HAL_NVIC_ClearPendingIRQ(context->irq);
```

Comment:
Here do:

```c
if (((isrflags & USART_SR_RXNE) != RESET) && ((cr1its & USART_CR1_RXNEIE) != RESET)) {
  our_own_receive(context);
} else {
  HAL_UART_IRQHandler(&context->handle);
}
```

Here is how to read the data from the peripheral: https://github.com/adafruit/stm32f4/blob/89a356f0b44883d55c45a5d22940777a44a2b881/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_uart.c#L2976-L2999

Reply:
@tannewt @hierophect and I had a side conversation about using HAL vs not. Assuming the current code nowworks, I said let's leave it alone for now and revisit this later, in order to move on to other things.